src/records.rs

```diff
@@ -0,0 +1,303 @@
+        &mut self.record
+    }
+}
+
+type RecordsStore = KeyValueStore<u64, Record>;
+
+pub struct Records {
+    store: RecordsStore,
+    /// Record (size, uid) indexed by asset type, owner, and freeze status, for easy allocation as
+    /// transfer or freeze inputs. The records for each asset are ordered by increasing size, which
+    /// makes it easy to implement a worst-fit allocator that minimizes fragmentation.
+    asset_records:
+        PersistableHashMapBTreeMultiMap<(AssetCode, UserAddress, FreezeFlag), (RecordAmount, u64)>,
+    /// Record uids indexed by nullifier, for easy removal when confirmed as transfer inputs
+    nullifier_records: PersistableHashMap<Nullifier, u64>,
+}
+
+impl Records {
+    #![allow(dead_code)]
+
+    pub fn new<L: Ledger>(
+        loader: &mut AtomicStoreLoader,
+        adaptor: EncryptingResourceAdapter<(u64, Option<Record>)>,
+        fill_size: u64,
+    ) -> Result<Self, KeystoreError<L>> {
+        let log = AppendLog::load(loader, adaptor, "keystore_records", fill_size)?;
+        let store = RecordsStore::new(log)?;
+        let mut records = Self {
+            store,
+            asset_records: Persistable::new(),
+            nullifier_records: Persistable::new(),
+        };
+        records.reload();
+        Ok(records)
+    }
+
+    /// Create a Record
+    ///
+    /// Returns an editor to the newly created Record
+    pub fn create<L: Ledger>(
+        &mut self,
+        uid: u64,
+        ro: RecordOpening,
+        nullifier: Nullifier,
+        hold_until: Option<u64>,
+    ) -> Result<RecordEditor<'_>, KeystoreError<L>> {
+        let record = Record {
+            ro,
+            uid,
+            nullifier,
+            hold_until,
+        };
+        let ro = &record.ro;
+        self.asset_records.insert((
+            (ro.asset_def.code, ro.pub_key.address(), ro.freeze_flag),
+            (record.ro.amount.into(), record.uid),
+        ));
+        self.nullifier_records
+            .insert((record.nullifier, record.uid));
+        let mut editor = RecordEditor::new(self, record);
+        editor.save()?;
+        Ok(editor)
+    }
+
+    /// Rebuild the indices for this store
+    fn reload(&mut self) {
+        self.asset_records = Persistable::new();
+        self.nullifier_records = Persistable::new();
+        for record in self.store.iter() {
+            let ro = &record.ro;
+            self.asset_records.insert((
+                (
+                    ro.asset_def.code,
+                    ro.pub_key.address().clone(),
+                    ro.freeze_flag,
+                ),
+                (record.ro.amount.into(), record.uid),
+            ));
+            self.nullifier_records
+                .insert((record.nullifier, record.uid));
+        }
+    }
+
+    fn store<L: Ledger>(&mut self, uid: u64, record: &Record) -> Result<(), KeystoreError<L>> {
+        self.store.store(&uid, record)?;
+        self.reload();
```

Comment:
Since the keys (uid, nullifier, asset, frozen, owner) are immutable, do we have to reload here? We may be able to keep the indices in sync more cheaply by just inserting/removing appropriately on every `create`/`delete` operation

Reply:
We don't I was actually just looking at the reload() stuff, I think it should not have to be used anywhere except when we create the store object.  But can probably just inline it there